griptape/events/event_bus.py

```diff
@@ -1,24 +1,28 @@
 from __future__ import annotations
 
-import threading
+import contextvars
 from typing import TYPE_CHECKING
 
-from attrs import Factory, define, field
+from attrs import define
 
 from griptape.mixins.singleton_mixin import SingletonMixin
 
 if TYPE_CHECKING:
     from griptape.events import BaseEvent, EventListener
 
 
+_event_listeners: contextvars.ContextVar[list[EventListener]] = contextvars.ContextVar("event_listeners", default=[])
+
+
 @define
 class _EventBus(SingletonMixin):
-    _event_listeners: list[EventListener] = field(factory=list, kw_only=True, alias="_event_listeners")
-    _thread_lock: threading.Lock = field(default=Factory(lambda: threading.Lock()), alias="_thread_lock")
-
     @property
     def event_listeners(self) -> list[EventListener]:
-        return self._event_listeners
+        return _event_listeners.get()
+
+    @event_listeners.setter
+    def event_listeners(self, event_listeners: list[EventListener]) -> None:
+        _event_listeners.set(event_listeners)
```

Comment:
Just for my own understanding, what would be the difference between setting the whole list versus updating the contents?

In other words what if this looked like:
```py
self.event_listeners.clear()
self.event_listeners.extend(event_listeners)
```

Like what if some class previously stored a reference to the return value from `EventBus.event_listeners`, then the setter was called? If you set the whole list then the stored reference will contain event listeners no longer in the event bus, is that the desired behavior?

Is there a restriction on mutating the value returned by `_event_listeners.get()` because of `contextvars`?

Reply:
The important thing is that `_event_listeners.set` is called at some point. While we could probably keep the in-place modifications like `append/remove/clear`, we'd need to call `set` later in order for the context variable to appear in `copy_context`. I can add a comment that explains this behavior.

Reply:
Thanks for the comment! Giving 🐑 , but will want to discuss offline